sql-odbc/src/PowerBIConnector/OdfeSqlOdbcPBIConnector.pq

```diff
@@ -1,42 +1,203 @@
+                    in
+                        [
+                            Text = Text.Format("LIMIT #{0}#{1}", { skip, take }),
+                            Location = "AfterQuerySpecification"
+                        ]
+        ],
+
+        OdbcDatasource = Odbc.DataSource(ConnectionString & Server & CredentialConnectionString, [
+            // Do not view the tables grouped by their schema names.
+            HierarchicalNavigation = false,
+            // Prevents execution of native SQL statements. Extensions should set this to true.
+            HideNativeQuery = true,
+            // Allows upconversion of numeric types
+            SoftNumbers = true,
+            // Allow upconversion / resizing of numeric and string types
+            TolerateConcatOverflow = true,
+            // Enables connection pooling via the system ODBC manager
+            ClientConnectionPooling = true,
+
+            // These values should be set by previous steps
+            AstVisitor = AstVisitor,
+            SQLColumns = SQLColumns,
+            SQLGetTypeInfo = SQLGetTypeInfo,
+
+            OnError = OnOdbcError,
+
+            // Connection string properties used for encrypted connections.
+            CredentialConnectionString = EncryptedConnectionString
+        ])
     in
         OdbcDatasource;
 
+// Handles ODBC errors.
+OnOdbcError = (errorRecord as record) =>
+    let
+        ErrorMessage = errorRecord[Message],
+        ConnectionHostPort = errorRecord[Detail][DataSourcePath], 
+        
+        IsDriverNotInstalled = Text.Contains(ErrorMessage, "doesn't correspond to an installed ODBC driver"),
+        
+        OdbcError = errorRecord[Detail][OdbcErrors]{0},
+        OdbcErrorCode = OdbcError[NativeError],
+        
+        // Failed to connect to given host
+        IsHostUnreachable = 
+            OdbcErrorCode = 202
+    in
+        if IsDriverNotInstalled then
+            error Error.Record("", "The Open Distro For Elasticsearch SQL ODBC driver is not installed. Please install the driver")
+        else if IsHostUnreachable then 
+            error Error.Record("", "Couldn't reach server. Please double-check the host, port and auth.")
+        else 
+            error errorRecord;
+
 // Data Source Kind description
 OdfeSqlOdbcPBIConnector = [
+    // Required for use with Power BI Service.
+    TestConnection = (dataSourcePath) => 
+        let
+            json = Json.Document(dataSourcePath),
+            Host = json[Host],
+            Port = json[Port]
+        in
+            { "OdfeSqlOdbcPBIConnector.Contents", Host, Port },
+
+    // Authentication modes
     Authentication = [
-        Implicit = []
+        Implicit = [
+            Label = "NONE"
+        ],
+        UsernamePassword = [
+            Label = "BASIC"
+        ],
+        Key = [
+            Label = "AWS_SIGV4",
+            KeyLabel = "Region"
+        ]
     ],
+
+    // Enable Encryption
+    SupportsEncryption = true,
```

Comment:
Note that if you set this to true, PBI will always try encrypted connections. You need to add an error handler that tells PBI to try an unencrypted connection if it gets an SSL error. See the ImpalaODBC example on the PowerQuery SDK github.

Reply:
Yes, will include that in separate PR since some changes are required in driver code to return proper error code for SSL.